website/addons/dataverse/views.py

```diff
@@ -0,0 +1,332 @@
+        )
+        provider.account.save()
+    except KeyExistsException:
+        # ... or get the old one
+        provider.account = ExternalAccount.find_one(
+            Q('provider', 'eq', provider.short_name) &
+            Q('provider_id', 'eq', api_token)
+        )
+        assert provider.account is not None
+
+    if provider.account not in user.external_accounts:
+        user.external_accounts.append(provider.account)
+
+    user_addon = auth.user.get_addon('dataverse')
+    if not user_addon:
+        user.add_addon('dataverse')
+    user.save()
+
+    # Need to ensure that the user has dataverse enabled at this point
+    user.get_or_add_addon('dataverse', auth=auth)
+    user.save()
+
+    return {}
+
+@must_have_permission('write')
+@must_have_addon(SHORT_NAME, 'user')
+@must_have_addon(SHORT_NAME, 'node')
+@must_be_addon_authorizer(SHORT_NAME)
+def dataverse_set_config(node_addon, auth, **kwargs):
+    """Saves selected Dataverse and dataset to node settings"""
+
+    user_settings = node_addon.user_settings
+    user = auth.user
+
+    if user_settings and user_settings.owner != user:
+        raise HTTPError(http.FORBIDDEN)
+
+    try:
+        assert_clean(request.json)
+    except AssertionError:
+        # TODO: Test me!
+        raise HTTPError(http.NOT_ACCEPTABLE)
+
+    alias = request.json.get('dataverse').get('alias')
+    doi = request.json.get('dataset').get('doi')
+
+    if doi is None:
+        return HTTPError(http.BAD_REQUEST)
+
+    connection = client.connect_from_settings(node_addon)
+    dataverse = client.get_dataverse(connection, alias)
+    dataset = client.get_dataset(dataverse, doi)
+
+    node_addon.set_folder(dataverse, dataset, auth)
+
+    return {'dataverse': dataverse.title, 'dataset': dataset.title}, http.OK
+
+
+@must_have_permission('write')
+@must_have_addon(SHORT_NAME, 'user')
+@must_have_addon(SHORT_NAME, 'node')
+def dataverse_get_datasets(node_addon, **kwargs):
+    """Get list of datasets from provided Dataverse alias"""
+    alias = request.json.get('alias')
+
+    connection = client.connect_from_settings(node_addon)
+    dataverse = client.get_dataverse(connection, alias)
+    datasets = client.get_datasets(dataverse)
+    ret = {
+        'alias': alias,  # include alias to verify dataset container
+        'datasets': [{'title': dataset.title, 'doi': dataset.doi} for dataset in datasets],
+    }
+    return ret, http.OK
+
+## Crud ##
+
+
+@must_have_permission('write')
+@must_not_be_registration
+@must_have_addon(SHORT_NAME, 'node')
+@must_be_addon_authorizer(SHORT_NAME)
+def dataverse_publish_dataset(node_addon, auth, **kwargs):
+    node = node_addon.owner
+    publish_both = request.json.get('publish_both', False)
+
+    now = datetime.datetime.utcnow()
+
+    try:
+        connection = client.connect_from_settings_or_401(node_addon)
+    except HTTPError as error:
+        if error.code == http.UNAUTHORIZED:
+            connection = None
```

Comment:
It's possible for an `HTTPError(http.SERVICE_UNAVAILABLE)` to trickle up here if there's a problem with the API, rather than with credentials... That being said, it does seem like we should be bailing out here if `connection == None`.

(These view functions were mostly moved but unmodified from `/dataverse/views/<file>.py`. Many of those were replaced with `generic_views`, but they're all stored in the same file, now.)